### PR TITLE
Declarative containers: more veth interfaces

### PR DESCRIPTION
**Pull request by kampfschlaefer**
With these changes, a container can have more then one veth-pair. This allows for example to have LAN and DMZ as bridges on the host and add dedicated containers for proxies, ipv4-firewall and ipv6-firewall. Or to have a bridge for normal WAN, one bridge for administration and one bridge for customer-internal communication. So that web-server containers can be reached from outside per http, from the management via ssh and can talk to their database via the customer network.

The scripts to set up the containers are now rendered several times instead of just one template. The scripts now contain per-container code to configure the extra veth interfaces. The default template without support for extra-veths is still rendered for the imperative containers.

Also a test is there to see if extra veths can be placed into host-bridges or can be reached via routing.

cc @edolstra @fpletz 

---
###### Things done
- Built on platform(s)
  - [x] NixOS
  - [ ] OS X
  - [x] Linux
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by kampfschlaefer** [comment]
I would rebase this PR to the current master (and have done so locally) with its updated systemd, but current master has broken containers…

File: nixos/modules/virtualisation/containers.nix [hunk 1]
```diff
@@ -22,37 +22,224 @@ let
...
-  containerInit = pkgs.writeScript "container-init"
-    ''
-      #! ${pkgs.stdenv.shell} -e
+  containerInit = (cfg:
+    let
+      renderExtraVeth = (name: cfg:
+        ''
+        echo "Bringing ${name} up"
+        ip link set dev ${name} up
+        ${if cfg . "localAddress" or null != null then
```
Comment (joachifm) [inline nixos/modules/virtualisation/containers.nix]: `if condition then "foo" else ""` is more neatly expressed as `optionalString condition "foo"`.

File: nixos/tests/containers-extra_veth.nix [hunk 2]
```diff
@@ -0,0 +1,103 @@
+# Test for NixOS' container support.
```
Comment (joachifm) [inline nixos/tests/containers-extra_veth.nix]: I might be missing something, but I can't see where this is added to the release tests?

**Comment by joachifm** [comment]
@fpletz anything we can do to help move this along?

File: nixos/tests/containers-extra_veth.nix (hunk 2 above)
Comment (kampfschlaefer) [inline nixos/tests/containers-extra_veth.nix]: @joachifm 
> I might be missing something, but I can't see where this is added to the release tests?

Argh, I missed this (again). Will add it shortly.

**Comment by kampfschlaefer** [comment]
@fpletz ping?

Would be nice to get that into master before 16.09 is branched off…

**Comment by fpletz** [comment]
Sorry for the long wait. I just had another look, let's merge this! :+1:

**Pull request merged**